### ISSUE TITLE
Save working directory for commands

### ISSUE THREAD

**Issue by notramo** [elves/elvish#673]
The command history should save working directory also.

**Comment by krader1961** [comment]
While I think this proposal is probably a good idea I'd like to hear a concrete proposal (or two) for how the information might be used. Other than simply displaying the info when running something like a hypothetical `hist list` command. There should be a compelling use case to justify recording the CWD with each command.

Note: This feature has been requested many times in the fish shell project, going back at least eight years and, so far as I'm aware, has yet to be implemented (see https://github.com/fish-shell/fish-shell/issues/120 as well as many others). The only concrete example in the fish project for how this information might be used is a mention of this zsh module: https://github.com/jimhester/per-directory-history. And that is actually a very different animal than including the CWD/PWD in effect when an interactive command is run.

**Comment by iwoloschin** [comment]
As a counterpoint, how hard would it be to implement? I tend to take a "record everything!" mentality because storage is cheap now and I don't know how I'll want to use data from today when I'm looking back at it tomorrow.

For a concrete proposal, I think the biggest thing for me is something like `docker`. I have a bunch of docker based projects, so there is a lot of `docker build ...` and `docker run ...` flying around in my various terminals. I tend to leave my terminals open on a specific project for days so I can often have an impromptu "per directory history" since Elvish's history does not fast-forward by default (yay!). However, if I close the terminal for any reason, now history is "jumbled" and a simple `docker b<up arrow>` will not necessarily find the correct command. It's trivial to just keep pressing the up arrow until I find it, but I find this catches me off guard more than I care to admit.

Instead, it'd be neat if Elvish could use `$pwd` to suggest the most recent command run *in that directory* first. After exhausting `$pwd`'s commands it'd just do most recently used for all history (ignoring the history's working directory). This could be jarring or unusual to new users, but I think this is also something that could feel intuitive and "smart". Or maybe I'm just crazy :).

**Comment by zzamboni** [comment]
I agree with @iwoloschin, it would be possible also to bind a different key (e.g. `Alt-Up`) to "prev matching command in the same directory". This could be quite useful.

**Comment by krader1961** [comment]
How would this interact with [ctrl-R]? Should there be a global (as opposed to mode specific) toggle that affects whether displayed history entries are limited to those run in the CWD, or at least present those entries ahead of all others? I don't think I'd want that to always be the default, or only, behavior when navigating command history.

Note that fish does not record the CWD but does record the pathnames present in each command. When presenting an autosuggestion it takes into account whether those pathnames are present in the CWD to affect the score assigned to each history entry and thus how likely it is to be shown as the autosuggested command. I remember that the first week I used fish (after having used ksh, then zsh for many years) I found that behavior disconcerting. And then I grew accustomed to it and wondered why every shell doesn't do something similar :smile:
